extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/worker/WorkerStagePhase.java

```diff
@@ -54,11 +54,12 @@ public boolean canTransitionFrom(final WorkerStagePhase priorPhase)
     @Override
     public boolean canTransitionFrom(final WorkerStagePhase priorPhase)
     {
-      return priorPhase == PRESHUFFLE_WAITING_FOR_RESULT_PARTITION_BOUNDARIES;
+      return priorPhase == PRESHUFFLE_WAITING_FOR_RESULT_PARTITION_BOUNDARIES /* if globally sorting */
+             || priorPhase == READING_INPUT /* if locally sorting */;
```

Comment:
curious why this never got caught before this patch?

Reply:
Before this patch, stages that locally sort would not enter the `PRESHUFFLE_WRITING_OUTPUT` phase. They would go directly from `READING_INPUT` to `RESULTS_READY`. But now, they need to enter `PRESHUFFLE_WRITING_OUTPUT` so they can send `doneReadingInput` to the controller.